### DISCUSSION TITLE
Places to host subscription link

### DISCUSSION THREAD

**Discussion by its0ka** [XTLS/Xray-core#2540] [Q&A] [answered]
I want to host a subscription file and think of github, but everyone could look at it and sometimes github doesn't work in my country. And it's too risky to host it inside country on a VPS. Please send websites/hostings where i can host and update my subscription

**Comment by chise0713** [comment]
If cloudflare is working fine in your country, I suggest you use it to reverse proxy `raw.githubusercontent.com/UserName/Repository` .
1. You need to get a [presonal access token](https://github.com/settings/tokens?type=beta) , only give it read permission of your subscrpition repository.
2. Use this `_worker.js` at cloudflare worker, remember to add `API_KEY` variable in cloudflare dash board , and change URL path in the js.

```js
addEventListener('fetch', event => {
  event.respondWith(handleRequest(event.request));
});

async function handleRequest(request) {
  const url = new URL(request.url);
  const actualUrlStr = url.pathname.replace("/sub/","/")   // Change this to fit yours
  const githubRawURL = `https://raw.githubusercontent.com/UserName/Repository/master${actualUrlStr}`;

  const response = await fetchWithToken(githubRawURL, API_KEY); // Add API_KEY variable in cloudflare dash board
  const newResponse = new Response(response.body, response);

  newResponse.headers.set('Access-Control-Allow-Origin', '*');

  return newResponse;
}

async function fetchWithToken(url, token) {
  const headers = {
    'Authorization': `Bearer ${token}`
  };

  const response = await fetch(url, { headers });

  return response;
}
```
Then visit ` yourdomain.com/sub/PathToYourFile ` , cloudflare can use your `API_KEY` to access your private repository and send your subscrpition to you. 
Because HTTPS encrypted your URL path, no one can see trough it, so don't worry about your subscrpition will leak.

**Reply by its0ka** [reply]
my country has sni whitelist on all cloudflare ips, i can't connect with my own domain (tcp reset)

**Reply by chise0713** [reply]
Try use Xray's `dokodemo-door` on your vps to reverse proxy it ? Or try to use worker's domain, every worker will have a `*.workers.dev` domain bind to it.

**Reply by its0ka** [reply]
i don't get why would i want to proxy cloudflare through my vps if i can host everything on my vps without cloudflare.
i don't want to expose my vps ip in subscription.
i found working website hosting and uploaded subscription there and i update it with ftp

**Reply by chise0713** [reply]
Anyway, your VPS's IP will be exposed regardless, so you might as well do more tasks on it.
If you just want some way to access your subscrpition file, vercel is also a good choice.

**Reply by us254** [reply, marked as answer]
**Google Drive Direct Download**
https://drive.google.com/file/d/1-aja5ydKFWb5DjjrwK-HefWgxN8PL4CI/view?usp=share_link

**To convert the Google Drive link to a direct download link by modifying the URL, please follow these steps:**

1:  Copy the file ID from the original link. In this case, the file ID is "**1-aja5ydKFWb5DjjrwK-HefWgxN8PL4CI**".
2:  Open a new tab in your web browser and enter the following URL:
3:  https://drive.google.com/uc?export=download&id=
4:  Paste the file ID at the end of the URL, immediately after "**id=**". The full URL should look like this:
5:  https://drive.google.com/uc?export=download&id=1-aja5ydKFWb5DjjrwK-HefWgxN8PL4CI
6:  Press Enter to load the modified URL. The file should now start downloading immediately.

**Reply by its0ka** [reply]
thank you, it works nicely!

**Reply by its0ka** [reply]
but idk how to upload/replace the subscription from my linux server to gdrive, any tips?

**Comment by us254** [comment]
https://developers.google.com/drive/api/guides/manage-uploads#python_1

, you can use the Python Google Drive API client to upload a file to Google Drive. 

1. **Install the Google Client Library**: 
   Run the following command in your terminal to install the library:
   ```
   pip install --upgrade google-api-python-client google-auth-httplib2 google-auth-oauthlib
   ```

2. **Set up Google Drive API**:
   - Go to the Google Cloud Console (https://console.cloud.google.com/) and create a new project.
   - Enable the Google Drive API for your project.
   - Create credentials for the Drive API. This will give you a `client_id` and `client_secret`.
   - Download the JSON file of your credentials.

3. **Set up OAuth 2.0**:
   - Install the `oauth2client` library using pip: `pip install oauth2client`
   - Use the `client_id` and `client_secret` from your Google project to authorize your app.

4. **Upload a File**:
   Use the following Python code to upload a file to Google Drive:

```python
import google.auth
from googleapiclient.discovery import build
from googleapiclient.http import MediaFileUpload
from googleapiclient.errors import HttpError

def upload_basic():
    """Insert new file.
    Returns : Id's of the file uploaded

    Load pre-authorized user credentials from the environment.
    """
    creds, _ = google.auth.default()

    try:
        # create drive api client
        service = build('drive', 'v3', credentials=creds)

        file_metadata = {'name': 'example.txt'}
        media = MediaFileUpload('/usr/local/bin/example.txt',
                                mimetype='text/plain')
        
        file = service.files().create(body=file_metadata, media_body=media,
                                      fields='id').execute()
        print(F'File ID: {file.get("id")}')

    except HttpError as error:
        print(F'An error occurred: {error}')
        
upload_basic()
```

In this Python script, replace `'example.txt'` with the name you want for the file on Google Drive and `'/usr/local/bin/example.txt'` with the path to your file on your Ubuntu server. 

 You can set up credentials in your environment using the `export` command in your terminal:

```bash
export GOOGLE_APPLICATION_CREDENTIALS="/path/to/your/credentials.json"
```
Replace `"/path/to/your/credentials.json"` with the path where you stored the JSON file you downloaded when you created your credentials.

Then, run your Python script. If everything is set up correctly, the ID of the uploaded file should be printed in your terminal.

Remember to replace necessary placeholders with your actual values. File upload success depends on the correct setup of Google API credentials and correct file path on your Ubuntu server.

**Reply by its0ka** [reply]
thanks a lot!
